### PR TITLE
fix `atomicAllInc()`

### PR DESCRIPTION
**Pull request by psychocoderHPC**
The implementation of `atomicAllInc()` is broken for Volta and newer CUDA GPUs. CUDA warps are not implicit synchroniced and using warp functions can have side effects if we assume that asking for the active threads mask `activemask()` twice will result into the same result.

This bug exists in since 7 years or more in PIConGPU. I only observed problems if I compile in debug mode but in general the release mode should be affected too.

Possible effects of the bug:
- particles silently get deleted and other particles get cloned
- invalid memory access
- maybe it is also possible that we create particles out of the thin air (this is not fully clear)

### CODE REVIEW

File: include/pmacc/kernel/atomic.hpp [hunk 1]
```diff
@@ -78,13 +78,18 @@ namespace pmacc
...
              *
+             * For modern architectures see https://developer.nvidia.com/blog/using-cuda-warp-level-primitives/ *
              */
             template<typename T_Type>
             struct AtomicAllIncKepler<T_Type, true>
             {
                 template<typename T_Acc, typename T_Hierarchy>
                 HDINLINE T_Type operator()(const T_Acc& acc, T_Type* ptr, const T_Hierarchy& hierarchy)
                 {
+                    /* @attention mask must be used in any warp operation whcih supports a mask.
```
Comment (chillenzer) [inline include/pmacc/kernel/atomic.hpp]:
```suggestion
                    /* @attention mask must be used in any warp operation which supports a mask.
```

File: include/pmacc/kernel/atomic.hpp [hunk 2]
```diff
@@ -78,13 +78,18 @@ namespace pmacc
...
+                     * On CUDA calling activemask again could result into different results because warps are not
```
Comment (chillenzer) [inline include/pmacc/kernel/atomic.hpp]:
```suggestion
                     * On CUDA calling activemask again could result in different results because warps are not
```

File: include/pmacc/kernel/atomic.hpp [hunk 3]
```diff
@@ -78,13 +78,18 @@ namespace pmacc
...
+                     * implicit synchronized. This is different to HIP and old CUDA GPUs before Volat.
```
Comment (chillenzer) [inline include/pmacc/kernel/atomic.hpp]:
```suggestion
                     * implicitly synchronized. This is different to HIP and old CUDA GPUs before Volta.
```

File: include/pmacc/kernel/warp.hpp [hunk 4]
```diff
@@ -53,66 +53,70 @@ namespace pmacc
...
 
-        //! Broadcast a 32bit unsigned int
-        DINLINE uint32_t warpBroadcast(const uint32_t data, const int32_t srcLaneId)
-        {
-            return static_cast<uint32_t>(warpBroadcast(static_cast<int32_t>(data), srcLaneId));
-        }
 
-        //! Broadcast a 64bit unsigned int
-        DINLINE uint64_cu warpBroadcast(const uint64_cu data, const int32_t srcLaneId)
+        /** fallback for not nativly supported types
```
Comment (chillenzer) [inline include/pmacc/kernel/warp.hpp]:
```suggestion
        /** fallback for not natively supported types
```

File: include/pmacc/kernel/warp.hpp [hunk 5]
```diff
@@ -53,66 +53,70 @@ namespace pmacc
...
+         *
+         * emulates the brotcast by repeating 32bit warp broatcasts
+         */
+        template<typename T_MaskType, typename T_DataType, typename T_Sfinae = void>
+        struct WarpBroatcast
```
Comment (chillenzer) [inline include/pmacc/kernel/warp.hpp]:
```suggestion
        struct WarpBroadcast
```

File: include/pmacc/kernel/warp.hpp [hunk 6]
```diff
@@ -53,66 +53,70 @@ namespace pmacc
...
+        };
+
+        /** CUDA and HIP native supported data types */
+        template<typename T_MaskType, typename T_DataType>
+        struct WarpBroatcast<
+            T_MaskType,
+            T_DataType,
+            std::enable_if_t<matchAnyType<
+                T_DataType,
+                int,
```
Comment (chillenzer) [inline include/pmacc/kernel/warp.hpp]: Could we give this list a name somehow? Haven't looked up the details but it would be much nicer to have something like
```c++
using NativelySupportedTypes = ???;
// ...
MatchAnyType<T_DataType, NativelySupportedTypes...>;
```

File: include/pmacc/kernel/warp.hpp [hunk 7]
```diff
@@ -53,66 +53,70 @@ namespace pmacc
...
-        DINLINE uint32_t warpBroadcast(const uint32_t data, const int32_t srcLaneId)
-        {
-            return static_cast<uint32_t>(warpBroadcast(static_cast<int32_t>(data), srcLaneId));
-        }
 
-        //! Broadcast a 64bit unsigned int
-        DINLINE uint64_cu warpBroadcast(const uint64_cu data, const int32_t srcLaneId)
+        /** fallback for not nativly supported types
+         *
+         * emulates the brotcast by repeating 32bit warp broatcasts
```
Comment (chillenzer) [inline include/pmacc/kernel/warp.hpp]:
```suggestion
         * emulates the broadcast by repeating 32bit warp broadcasts
```

File: include/pmacc/kernel/warp.hpp [hunk 8]
```diff
@@ -53,66 +53,70 @@ namespace pmacc
...
 
-        //! Broadcast a 32bit float
-        DINLINE float warpBroadcast(const float data, const int32_t srcLaneId)
+                return data;
+            }
+        };
+
+        /** CUDA and HIP native supported data types */
+        template<typename T_MaskType, typename T_DataType>
+        struct WarpBroatcast<
```
Comment (chillenzer) [inline include/pmacc/kernel/warp.hpp]:
```suggestion
        struct WarpBroadcast<
```

File: include/pmacc/kernel/warp.hpp [hunk 9]
```diff
@@ -53,66 +53,70 @@ namespace pmacc
...
-        //! Broadcast a 64bit float by using 2 32bit broadcasts
-        DINLINE double warpBroadcast(double data, const int32_t srcLaneId)
+        template<typename T_MaskType, typename T_DataType>
+        DINLINE T_DataType warpBroadcast(T_MaskType const mask, T_DataType const data, int const srcLaneId)
         {
-            float* const pData = reinterpret_cast<float*>(&data);
-            pData[0] = warpBroadcast(pData[0], srcLaneId);
-            pData[1] = warpBroadcast(pData[1], srcLaneId);
-            return data;
+            return WarpBroatcast<T_MaskType, T_DataType>{}(mask, data, srcLaneId);
```
Comment (chillenzer) [inline include/pmacc/kernel/warp.hpp]:
```suggestion
            return WarpBroadcast<T_MaskType, T_DataType>{}(mask, data, srcLaneId);
```

Comment (chillenzer) [inline include/pmacc/kernel/warp.hpp]: Can't we rely on type deduction here?

File: include/pmacc/kernel/warp.hpp [hunk 10]
```diff
@@ -53,66 +53,70 @@ namespace pmacc
...
+         */
+        template<typename T_MaskType, typename T_DataType, typename T_Sfinae = void>
+        struct WarpBroatcast
         {
-            return static_cast<uint64_cu>(warpBroadcast(static_cast<int64_cu>(data), srcLaneId));
-        }
+            DINLINE T_DataType operator()(T_MaskType const mask, T_DataType data, int const srcLaneId) const
+            {
+                static_assert(sizeof(T_DataType) >= sizeof(int) && (sizeof(T_DataType) % sizeof(int)) == 0);
+                int* const pData = reinterpret_cast<int*>(&data);
```
Comment (chillenzer) [inline include/pmacc/kernel/warp.hpp:89]: The previous implementation was more precise about the type here: `sizeof(int)` is not necessarily 32bit. Should we use `int32_t` here as before?

File: include/pmacc/kernel/warp.hpp [hunk 11]
```diff
@@ -53,66 +53,70 @@ namespace pmacc
...
-        /** broadcast data within a warp
-         *
-         * required PTX ISA >=3.0
+
+        /** broadcast data within a warp without using shared memory
          *
+         * @param mask mask to select threads participating in the operation
          * @param data value to broadcast
          * @param srcLaneId lane id of the source thread
-         * @return value send by the source thread
```
Comment (chillenzer) [inline include/pmacc/kernel/warp.hpp:62]: Why is this deleted?

File: include/pmacc/kernel/warp.hpp [hunk 12]
```diff
@@ -53,66 +53,70 @@ namespace pmacc
 
 
 #    if(__CUDA_ARCH__ >= 300 || BOOST_COMP_HIP)
-        /** broadcast data within a warp
-         *
-         * required PTX ISA >=3.0
```
Comment (chillenzer) [inline include/pmacc/kernel/warp.hpp:58]: Was this requirement dropped?

**Review by chillenzer — CHANGES_REQUESTED** [review]
Hi, I could have waited for the CI but somehow I didn't. Looks good so far. Very nice how you abstracted away all the different special cases into one elegant template.

My comments are mostly typos (and I think GH has a button for immediately accepting the suggestions in simple cases) but I think I've found a mismatch of signatures that we should check. Might have overlooked something though.

File: include/pmacc/kernel/warp.hpp (hunk 6 above)
Comment (psychocoderHPC) [reply include/pmacc/kernel/warp.hpp]: This would be nice but would make the generator expression more complicated because I need the variable template signature. ONly a workaround with `std::tuple` and unfolding the tuple with `std::Get<>` would come into my mind, but IMO the expression will be much more complicated to read.

File: include/pmacc/kernel/warp.hpp (hunk 10 above)
Comment (psychocoderHPC) [reply include/pmacc/kernel/warp.hpp:89]: I changed it back to `int` because the native interfaces requires `int` and in principle `int32_t` could be anything.
So the new implementation will guarantee that all works. The old most likely worked but  if we are pedantic than there is no guarantee.

File: include/pmacc/kernel/warp.hpp (hunk 12 above)
Comment (psychocoderHPC) [reply include/pmacc/kernel/warp.hpp:58]: All supported CUDA architectures are 3.5 anyway and even 3.5 is not supported by CUDA 12.X anymore.
So no need to mention it because it is supported for all GPUs (NVIDIA and AMD)

File: include/pmacc/kernel/warp.hpp (hunk 9 above)
Comment (psychocoderHPC) [reply include/pmacc/kernel/warp.hpp]: No this is not working because I do not forwarded the arguments to the constructor so the type can not be deducted.

**Comment by psychocoderHPC** [comment]
I applied @chillenzer comments. I hit a new record in typos per line of code^^

File: include/pmacc/kernel/warp.hpp (hunk 6 above)
Comment (chillenzer) [reply include/pmacc/kernel/warp.hpp]: It isn't too ugly:
```c++
#include <iostream>
#include <type_traits>

template <typename...> struct TypeList {};

using NativeTypes = TypeList<int, float>;

template <typename T, typename... T_TypeList,
          template <typename...> typename T_Pack>
inline constexpr bool matchAnyType(T_Pack<T_TypeList...>) {
  return (std::is_same_v<T, T_TypeList> || ...);
}

template <typename T, typename = void> struct F {
  F() { std::cout << "Called fallback!\n"; }
};

template <typename T>
struct F<T, std::enable_if_t<matchAnyType<T>(NativeTypes{})>> {
  F() { std::cout << "Called implementation!\n"; }
};

int main() {
  std::cout << matchAnyType<int>(NativeTypes{}) << "\n";
  std::cout << matchAnyType<float>(NativeTypes{}) << "\n";
  std::cout << matchAnyType<char>(NativeTypes{}) << "\n";

  F<int>();
  F<float>();
  F<char>();
}
```

File: include/pmacc/kernel/warp.hpp (hunk 9 above)
Comment (chillenzer) [reply include/pmacc/kernel/warp.hpp]: Okay. Admittedly, it is kind of the reason why we have these wrapper functions after all...^^

File: include/pmacc/kernel/warp.hpp (hunk 6 above)
Comment (psychocoderHPC) [reply include/pmacc/kernel/warp.hpp]: I will try https://godbolt.org/z/4nzG1eree based on your solution

Comment (chillenzer) [reply include/pmacc/kernel/warp.hpp]: Nice one! Didn't know about `std::disjunction`!

Comment (psychocoderHPC) [reply include/pmacc/kernel/warp.hpp]: applied the changes to the latest version of this PR

**Review by chillenzer — APPROVED** [review]

**Review by steindev — APPROVED** [review]
Relying on @chillenzer's review :sweat_smile:
